boot/flags.go

```diff
@@ -0,0 +1,251 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	errNotUC20 = fmt.Errorf("cannot get boot flags on non-UC20 device")
+
+	understoodBootFlags = []string{
+		// the factory boot flag is set to indicate that this is a
+		// boot inside a trusted factory environment
+		"factory",
+	}
+)
+
+// splitFlagString splits the given comma delimited list of boot flags, removing
+// empty strings.
+// Note that this explicitly does not filter out unsupported boot flags in the
+// off chance that an old version of the initramfs is reading new boot flags
+// written by a new version of snapd in userspace on a previous boot.
+func splitFlagString(s string) []string {
+	flags := []string{}
+	for _, flag := range strings.Split(s, ",") {
+		if flag != "" {
+			flags = append(flags, flag)
+		}
+	}
+
+	return flags
+}
+
+func checkFlagList(flags []string, allowList []string) error {
+	if len(allowList) != 0 {
+		// then we need to enforce the allow list
+		for _, flag := range flags {
+			if !strutil.ListContains(allowList, flag) {
+				return fmt.Errorf("flag %q is not allowed", flag)
+			}
+		}
+	}
+	return nil
+}
+
+func dropEmptyBootFlags(flags []string) []string {
+	// drop empty strings before serializing
+	nonEmptyFlags := make([]string, 0, len(flags))
+	for _, flag := range flags {
+		if strings.TrimSpace(flag) != "" {
+			nonEmptyFlags = append(nonEmptyFlags, flag)
+		}
+	}
+	return nonEmptyFlags
+}
+
+func serializeFlagString(flags []string) string {
+	return strings.Join(flags, ",")
+}
+
+// SetImageBootFlags writes the provided flags to the bootenv of the recovery
+// bootloader in the specified system rootDir. It is only meant to be called at
+// prepare-image customization time by ubuntu-image/prepare-image.
+func SetImageBootFlags(flags []string, rootDir string) error {
+	// check that the flagList is supported
+	if err := checkFlagList(flags, understoodBootFlags); err != nil {
+		return err
+	}
+
+	// also ensure that the serialized value of the boot flags fits inside the
+	// bootenv value, on lk systems the max size of a bootenv value is 255 chars
+	s := serializeFlagString(dropEmptyBootFlags(flags))
```

Comment:
afaict we don't need the dropEmptyBootFlags because checkFlagList would have errored already on them

Reply:
ah yes good point we would have